File: aulas_novas/Aula01/codigos/aula01.go
// =====================================================================
//
//	Paradigmas de Programação - Aula 1
//	Linguagem: Go        Plataforma online: https://go.dev/play
//	                     (cole o arquivo inteiro e clique em Run)
//	Go cobre CSP e canais na Aula 15.
//
// =====================================================================
package main

import (
	"fmt"
	"strings"
	"sync"
	"time"
)

const N = 200000

func dados() []int {
	xs := make([]int, N)
	for i := range xs {
		xs[i] = i + 1
	}
	return xs
}

// 1. IMPERATIVO: o estilo idiomático em Go. Sem map/filter na linguagem.
func imperativo(xs []int) int {
	total := 0
	for _, x := range xs {
		if x%2 == 0 {
			total += x * x // mutação explícita
		}
	}
	return total
}

// 2. ORDEM SUPERIOR: Go tem funções de primeira classe, ainda que verbosas
func filtrar(xs []int, p func(int) bool) []int {
	saida := make([]int, 0, len(xs))
	for _, x := range xs {
		if p(x) {
			saida = append(saida, x)
		}
	}
	return saida
}

func mapear(xs []int, f func(int) int) []int {
	saida := make([]int, len(xs))
	for i, x := range xs {
		saida[i] = f(x)
	}
	return saida
}

func reduzir(xs []int, inicial int, f func(int, int) int) int {
	acc := inicial
	for _, x := range xs {
		acc = f(acc, x)
	}
	return acc
}

func ordemSuperior(xs []int) int {
	pares := filtrar(xs, func(x int) bool { return x%2 == 0 })
	quadrados := mapear(pares, func(x int) int { return x * x })
	return reduzir(quadrados, 0, func(a, b int) int { return a + b })
}

//  3. CSP: goroutines e canais. O estado NÃO é compartilhado — ele viaja
//     pelo canal. Prévia da Aula 15.
func porCanais(xs []int, trabalhadores int) int {
	tamanho := len(xs) / trabalhadores
	parciais := make(chan int, trabalhadores)
	var wg sync.WaitGroup

	for w := 0; w < trabalhadores; w++ {
		inicio := w * tamanho
		fim := inicio + tamanho
		if w == trabalhadores-1 {
			fim = len(xs)
		}
		wg.Add(1)
		go func(fatia []int) { // goroutine: processo leve
			defer wg.Done()
			soma := 0
			for _, x := range fatia {
				if x%2 == 0 {
					soma += x * x
				}
			}
			parciais <- soma // comunicação por canal, não por memória
		}(xs[inicio:fim])
	}
	wg.Wait()
	close(parciais)

	total := 0
	for p := range parciais {
		total += p
	}
	return total
}

// 4. MEMÓRIA COMPARTILHADA COM LOCK: o outro modelo, o da Aula 14
func porMutex(xs []int, trabalhadores int) int {
	total := 0
	var mu sync.Mutex
	var wg sync.WaitGroup
	tamanho := len(xs) / trabalhadores

	for w := 0; w < trabalhadores; w++ {
		inicio := w * tamanho
		fim := inicio + tamanho
		if w == trabalhadores-1 {
			fim = len(xs)
		}
		wg.Add(1)
		go func(fatia []int) {
			defer wg.Done()
			soma := 0
			for _, x := range fatia {
				if x%2 == 0 {
					soma += x * x
				}
			}
			mu.Lock() // sem este lock, o resultado seria não determinístico
			total += soma
			mu.Unlock()
		}(xs[inicio:fim])
	}
	wg.Wait()
	return total
}

func barra(v, pico float64) string {
	n := int(v / pico * 40)
	if n < 1 {
		n = 1
	}
	return strings.Repeat("#", n)
}

func main() {
	xs := dados()
	esperado := imperativo(xs)

	fmt.Println(strings.Repeat("=", 62))
	fmt.Println("  PARADIGMAS DE PROGRAMACAO - AULA 1 - Go")
	fmt.Println("  Problema: somar os quadrados dos pares de 1 a 200.000")
	fmt.Println(strings.Repeat("=", 62))
	fmt.Printf("  Resultado esperado: %d\n\n", esperado)

	type medida struct {
		nome string
		ms   float64
	}
	versoes := []struct {
		nome string
		f    func() int
	}{
		{"Imperativo", func() int { return imperativo(xs) }},
		{"Ordem superior", func() int { return ordemSuperior(xs) }},
		{"CSP (4 canais)", func() int { return porCanais(xs, 4) }},
		{"Mutex (4 threads)", func() int { return porMutex(xs, 4) }},
	}

	medidas := []medida{}
	pico := 0.0
	for _, v := range versoes {
		t0 := time.Now()
		r := v.f()
		ms := float64(time.Since(t0).Microseconds()) / 1000
		if r != esperado {
			panic("resultado divergente em " + v.nome)
		}
		if ms > pico {
			pico = ms
		}
		medidas = append(medidas, medida{v.nome, ms})
	}

	// O sandbox do Go Playground usa um relogio FALSO, que so avanca quando
	// todas as goroutines bloqueiam. Por isso os tempos podem sair zerados la:
	// isso nao e um erro do programa, e uma propriedade do ambiente.
	if pico <= 0 {
		fmt.Println("  TEMPO DE EXECUCAO: nao mensuravel neste ambiente")
		fmt.Println("  (o Go Playground congela o relogio durante a computacao;")
		fmt.Println("   rode localmente com 'go run aula01.go' para ver os tempos)")
	} else {
		fmt.Println("  TEMPO DE EXECUCAO")
		for _, m := range medidas {
			fmt.Printf("  %-18s %7.2f ms  %s\n", m.nome, m.ms, barra(m.ms, pico))
		}
	}

	fmt.Println("\n  Os quatro devolvem o MESMO numero.")
	fmt.Println("  CSP e Mutex resolvem o mesmo problema de concorrencia por")
	fmt.Println("  caminhos opostos: mensagem versus memoria compartilhada.")
	fmt.Println("\n  >>> AMBIENTE GO OK <<<")
}

File: aulas_novas/Aula01/codigos/Aula01.java
// =====================================================================
//  Paradigmas de Programação - Aula 1
//  Linguagem: Java 17+   Plataforma online: https://onecompiler.com/java
//                        (ou JDoodle, Replit, OnlineGDB)
//  Java cobre orientação a objetos (Aula 11) e threads (Aula 14).
//  A classe precisa se chamar Aula01 e o arquivo, Aula01.java.
// =====================================================================
import java.util.List;
import java.util.ArrayList;
import java.util.stream.IntStream;
import java.util.stream.Collectors;

public class Aula01 {

    static final int N = 200_000;

    static List<Integer> dados() {
        List<Integer> xs = new ArrayList<>(N);
        for (int i = 1; i <= N; i++) xs.add(i);
        return xs;
    }

    // 1. IMPERATIVO: o estilo original da linguagem
    static long imperativo(List<Integer> xs) {
        long total = 0;
        for (int x : xs) {
            if (x % 2 == 0) total += (long) x * x;   // mutação
        }
        return total;
    }

    // 2. FUNCIONAL: streams, disponíveis desde o Java 8 (2014)
    static long funcional(List<Integer> xs) {
        return xs.stream()
                 .filter(x -> x % 2 == 0)
                 .mapToLong(x -> (long) x * x)
                 .sum();
    }

    // 3. ORIENTADO A OBJETOS: o estado encapsulado em um objeto
    static class Somador {
        private long total = 0;
        Somador oferecer(int x) {
            if (x % 2 == 0) total += (long) x * x;
            return this;
        }
        long total() { return total; }
    }

    static long orientadoAObjetos(List<Integer> xs) {
        Somador s = new Somador();
        for (int x : xs) s.oferecer(x);
        return s.total();
    }

    // 4. PARALELO: memória compartilhada gerenciada pela biblioteca.
    //    Só é seguro porque a operação é associativa e sem efeitos (Aula 14).
    static long paralelo(List<Integer> xs) {
        return xs.parallelStream()
                 .filter(x -> x % 2 == 0)
                 .mapToLong(x -> (long) x * x)
                 .sum();
    }

    static String barra(double v, double pico) {
        int n = Math.max(1, (int) Math.round(v / pico * 40));
        return "#".repeat(n);
    }

    public static void main(String[] args) {
        List<Integer> xs = dados();
        long esperado = imperativo(xs);

        System.out.println("=".repeat(62));
        System.out.println("  PARADIGMAS DE PROGRAMACAO - AULA 1 - Java");
        System.out.println("  Problema: somar os quadrados dos pares de 1 a 200.000");
        System.out.println("=".repeat(62));
        System.out.printf("  Resultado esperado: %d%n%n", esperado);

        String[] nomes = {"Imperativo", "Streams", "Orient. a objetos", "Paralelo"};
        double[] tempos = new double[4];

        for (int i = 0; i < 4; i++) {
            long t0 = System.nanoTime();
            long r = switch (i) {
                case 0 -> imperativo(xs);
                case 1 -> funcional(xs);
                case 2 -> orientadoAObjetos(xs);
                default -> paralelo(xs);
            };
            tempos[i] = (System.nanoTime() - t0) / 1_000_000.0;
            if (r != esperado) throw new IllegalStateException("divergiu em " + nomes[i]);
        }

        double pico = 0;
        for (double t : tempos) pico = Math.max(pico, t);

        System.out.println("  TEMPO DE EXECUCAO");
        for (int i = 0; i < 4; i++) {
            System.out.printf("  %-18s %7.2f ms  %s%n", nomes[i], tempos[i], barra(tempos[i], pico));
        }

        System.out.println();
        System.out.println("  Os quatro devolvem o MESMO numero.");
        System.out.println("  Java nasceu imperativa e OO; ganhou lambdas em 2014.");
        System.out.println("  Nenhuma linguagem viva fica num paradigma so (Aula 1).");
        System.out.println();
        System.out.println("  >>> AMBIENTE JAVA OK <<<");
    }
}
